### PR TITLE
feat: generic pages in contexts

### PR DESCRIPTION
**Pull request by vladfrangu**
TODO: responses too

### CODE REVIEW

File: packages/apify/src/crawlers/basic_crawler.ts [hunk 1]
```diff
@@ -678,7 +678,8 @@ export class BasicCrawler {
         }
     }
 
-    protected async _executeHooks(hooks: Hook[], ...args: Parameters<Hook>) {
+    // eslint-disable-next-line space-before-function-paren -- looks like a rule colliding with TS
```
Comment (B4nan) [inline packages/apify/src/crawlers/basic_crawler.ts:681]: yop https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/space-before-function-paren.md

Comment (B4nan) [reply packages/apify/src/crawlers/basic_crawler.ts:681]: should be fixed in master

File: packages/apify/src/crawlers/basic_crawler.ts [hunk 2]
```diff
@@ -678,7 +678,8 @@ export class BasicCrawler {
...
+    protected async _executeHooks<Hooks extends (...args: unknown[]) => Awaitable<void>>(hooks: Hooks[], ...args: Parameters<Hooks>) {
```
Comment (B4nan) [inline packages/apify/src/crawlers/basic_crawler.ts:682]: the type arg is called `Hooks` but it describes single hook, right?

File: packages/apify/src/crawlers/browser_crawler.ts [hunk 3]
```diff
@@ -17,17 +17,20 @@ import { BrowserLaunchContext } from '../browser_launchers/browser_launcher';
...
     GoToOptions extends Record<string, any> = Record<string, any>
-> = (crawlingContext: T, gotoOptions: GoToOptions) => Awaitable<void>;
-export type BrowserHook = Hook<BrowserCrawlingContext>;
-export type BrowserHandlePageFunction = (context: BrowserCrawlingContext) => Awaitable<void>;
-export type GotoFunction = (context: BrowserCrawlingContext, gotoOptions: Record<string, any>) => Awaitable<any>;
+> = (crawlingContext: Context, gotoOptions: GoToOptions) => Awaitable<void>;
+export type BrowserHandlePageFunction<Context = BrowserCrawlingContext> = (context: Context) => Awaitable<void>;
+export type GotoFunction<
+    Context = BrowserCrawlingContext,
+    GoToOptions extends Record<string, any> = Record<string, any>
```
Comment (B4nan) [inline packages/apify/src/crawlers/browser_crawler.ts:32]: can we use `Dictionary` here or do we need the keys to be `any`?

**Review by B4nan — APPROVED** [review]
good job, just few nitpicks

File: packages/apify/src/crawlers/basic_crawler.ts (hunk 2 above)
Comment (vladfrangu) [reply packages/apify/src/crawlers/basic_crawler.ts:682]: Yeah, I'll rename this type arg to `HookLike`

File: packages/apify/src/crawlers/cheerio_crawler.ts [hunk 4]
```diff
@@ -335,6 +335,7 @@ export interface CheerioHandlePageInputs extends CrawlingContext {
      */
     body: (string | Buffer);
 
+    // TODO: maybe consider making this `unknown` instead to force casting for TS users
```
Comment (B4nan) [inline packages/apify/src/crawlers/cheerio_crawler.ts:338]: why not make it another generic argument (and default to unknown)

**Review by B4nan — APPROVED** [review]

File: packages/apify/src/crawlers/cheerio_crawler.ts (hunk 4 above)
Comment (vladfrangu) [reply packages/apify/src/crawlers/cheerio_crawler.ts:338]: Can do that, sure

**Pull request merged**